### ISSUE TITLE
Link detection improvements

### ISSUE THREAD

**Issue by DeepDiver1975** [owncloud/mail#641]
Following links are not detected properly

* See <https://ci.owncloud.org/job/ownCloud-Documentation(7.0)/504/changes> (parenthesis)
* https://communities.coverity.com/community/scan-(open-source)/content – cuts off at the opening parenthesis.
* https://ma.ellak.gr/events/5ος-εκπαιδευτικός-κύκλος-σεμιναρίων-ε-5/

Also hashes # seem to cause problems.

**Comment by jancborchardt** [comment]
Moving to 0.2 – not a release blocker. ;)

**Comment by jancborchardt** [comment]
Same here: https://communities.coverity.com/community/scan-(open-source)/content – cuts off at the opening parenthesis.

**Comment by jancborchardt** [comment]
Anyone with regex-fu capable of fixing this? The code in question is at https://github.com/owncloud/mail/blob/master/lib/service/html.php#L99-L107

cc @clementhk @zinks- @irgendwie @wurstchristoph

**Comment by jancborchardt** [comment]
@LukasReschke get to work. :D

**Comment by kordianbruck** [comment]
In order to match unicode charaters one could use the category operator: http://www.regular-expressions.info/unicode.html#category

Also check out https://mathiasbynens.be/demo/url-regex 
Plenty of people tried to create a regex for this, no need to reinvent the wheel.

**Comment by clementhk** [comment]
Well, by a quick search on google, the one that passes all tests can be on https://mathiasbynens.be/demo/url-regex. But that one is for url validation, after a bit of modification it should work.

```
$regex = '_(?:(?:https?|ftp)://)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]-*)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]-*)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,})))(?::\d{2,5})?(?:/\S*)?_iuS';
```

**Comment by Xenopathic** [comment]
How about this? It will match any URI, as per the RFC spec (taken from http://stackoverflow.com/questions/30847/regex-to-validate-uris)

```
# RFC-3986 URI component:  URI
[A-Za-z][A-Za-z0-9+\-.]* :                                      # scheme ":"
(?: //                                                          # hier-part
  (?: (?:[A-Za-z0-9\-._~!$&'()*+,;=:]|%[0-9A-Fa-f]{2})* @)?
  (?:
    \[
    (?:
      (?:
        (?:                                                    (?:[0-9A-Fa-f]{1,4}:)    {6}
        |                                                   :: (?:[0-9A-Fa-f]{1,4}:)    {5}
        | (?:                            [0-9A-Fa-f]{1,4})? :: (?:[0-9A-Fa-f]{1,4}:)    {4}
        | (?: (?:[0-9A-Fa-f]{1,4}:){0,1} [0-9A-Fa-f]{1,4})? :: (?:[0-9A-Fa-f]{1,4}:)    {3}
        | (?: (?:[0-9A-Fa-f]{1,4}:){0,2} [0-9A-Fa-f]{1,4})? :: (?:[0-9A-Fa-f]{1,4}:)    {2}
        | (?: (?:[0-9A-Fa-f]{1,4}:){0,3} [0-9A-Fa-f]{1,4})? ::    [0-9A-Fa-f]{1,4}:
        | (?: (?:[0-9A-Fa-f]{1,4}:){0,4} [0-9A-Fa-f]{1,4})? ::
        ) (?:
            [0-9A-Fa-f]{1,4} : [0-9A-Fa-f]{1,4}
          | (?: (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?) \.){3}
                (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)
          )
      |   (?: (?:[0-9A-Fa-f]{1,4}:){0,5} [0-9A-Fa-f]{1,4})? ::    [0-9A-Fa-f]{1,4}
      |   (?: (?:[0-9A-Fa-f]{1,4}:){0,6} [0-9A-Fa-f]{1,4})? ::
      )
    | [Vv][0-9A-Fa-f]+\.[A-Za-z0-9\-._~!$&'()*+,;=:]+
    )
    \]
  | (?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}
       (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)
  | (?:[A-Za-z0-9\-._~!$&'()*+,;=]|%[0-9A-Fa-f]{2})*
  )
  (?: : [0-9]* )?
  (?:/ (?:[A-Za-z0-9\-._~!$&'()*+,;=:@]|%[0-9A-Fa-f]{2})* )*
| /
  (?:    (?:[A-Za-z0-9\-._~!$&'()*+,;=:@]|%[0-9A-Fa-f]{2})+
    (?:/ (?:[A-Za-z0-9\-._~!$&'()*+,;=:@]|%[0-9A-Fa-f]{2})* )*
  )?
|        (?:[A-Za-z0-9\-._~!$&'()*+,;=:@]|%[0-9A-Fa-f]{2})+
    (?:/ (?:[A-Za-z0-9\-._~!$&'()*+,;=:@]|%[0-9A-Fa-f]{2})* )*
|
)
(?:\? (?:[A-Za-z0-9\-._~!$&'()*+,;=:@/?]|%[0-9A-Fa-f]{2})* )?   # [ "?" query ]
(?:\# (?:[A-Za-z0-9\-._~!$&'()*+,;=:@/?]|%[0-9A-Fa-f]{2})* )?   # [ "#" fragment ]
```

:see_no_evil: :see_no_evil: :see_no_evil:

**Comment by Xenopathic** [comment]
Alternatively, we could move the function client-side with this handy JS library: http://soapbox.github.io/linkifyjs/

**Comment by jancborchardt** [comment]
cc @jcbrand on those regexes ^

@Xenopathic @clementhk also see the PR by @jcbrand at https://github.com/owncloud/mail/pull/797

**Comment by clementhk** [comment]
@jcbrand and guys, I think we should use something that is ready and well tested in other platforms and tests instead of reinventing.

**Comment by clementhk** [comment]
@Xenopathic wow that's a heavy lib for detecting links, respect to the author..

**Comment by clementhk** [comment]
If someone can find a better test than https://mathiasbynens.be/demo/url-regex then please post it so we can test it.

**Comment by jcbrand** [comment]
@Xenopathic Doesn't look to me like the Stackoverflow regex you show will match non-unicode characters. 
Linkify.js looks look a pretty cool solution though.

@clementhk The **diegoperini** regex matches all the urls in https://mathiasbynens.be/demo/url-regex, so that might be a good one to reuse.

**Comment by Xenopathic** [comment]
@jcbrand @jancborchardt I was joking - there's no way that beast can be considered maintainable. I agree with @clementhk here - we should be using a tried and tested solution rather than rolling our own (or pinching a regex from StackOverflow)

**Comment by DeepDiver1975** [comment]
fixed with #823
